### ISSUE TITLE
Socket connection from browser to custom microservice through webapp-starter

### ISSUE THREAD

**Issue by akshayeshenoi** [PredixDev/predix-webapp-starter#11]
Hi!
Is it currently possible to reverse-proxy WebSocket connections initiated by the browser to a custom microservice? The lack of it requires the front-end to connect to the microservice directly which entails authentication woes.

**Comment by gstroup** [comment]
No, currently the webapp-starter does not support proxying websocket connections.  It's a good idea, and something we've considered.  For now, you could look at the predix-toolkit code for an example of proxying web socket connections. 
https://github.com/PredixDev/predix-tool-kit/blob/master/server/security-app.js

Also, this npm package is very popular, and supports web socket proxying:
https://github.com/nodejitsu/node-http-proxy

**Comment by gstroup** [comment]
Yes, I think a PR would be useful.  How are you planning to implement it?

**Comment by akshayeshenoi** [comment]
Kind of like how it's implemented in the predix-toolkit example code you shared. 
- A websocket server will be instantiated on the webapp-starter.
- Clients will hit a certain endpoint to connect to the websocket on another microservice. 
- A new ws connection to the microservice will be established by the webapp-starter (acting as a client), and a new socketID will be generated for it which will be returned to the client. The socket generated will be stored in-memory against the ID.
- The client will then connect to the websocket we instantiated before and send the ID within the body of the message. The webapp-starter will use that ID to fetch the socket and forward the message.

All communication will depend upon this socketID. Sounds good?

**Comment by gstroup** [comment]
Sounds good.  Let's try to keep most of this logic in a separate file in the /server directory.  Thanks for the initiative.
